Core/SceneController.gd

```diff
@@ -0,0 +1,50 @@
+extends Node
+
+
+var current_scene = null
+
+func _ready():
+	var root = get_tree().root
+	current_scene = root.get_child(root.get_child_count() - 1)
+	PhaseManager.on_combat_end.connect(_combat_end_change_scene)
+
+
+func goto_scene(path):
+	# This function will usually be called from a signal callback,
+	# or some other function in the current scene.
+	# Deleting the current scene at this point is
+	# a bad idea, because it may still be executing code.
+	# This will result in a crash or unexpected behavior.
+
+	# The solution is to defer the load to a later time, when
+	# we can be sure that no code from the current scene is running:
+
+	call_deferred("_deferred_goto_scene", path)
+
+
+func _deferred_goto_scene(path):
+	# It is now safe to remove the current scene
+	current_scene.free()
+
+	# Load the new scene.
+	var s = ResourceLoader.load(path)
+
+	# Instance the new scene.
+	current_scene = s.instantiate()
+
+	# Add it to the active scene, as child of root.
+	get_tree().root.add_child(current_scene)
+
+	# Optionally, to make it compatible with the SceneTree.change_scene_to_file() API.
+	get_tree().current_scene = current_scene
+
+
+func _combat_end_change_scene(combat_result: Enums.CombatResult):
+	PhaseManager.call_deferred("set_phase", Enums.Phase.GAME_STARTING)
```

Comment:
So I'm guessing PhaseManager's `_ready` doesn't get called whenever we reload the scene which is why we need to call it manually here. Would it be problematic to move the logic from the `_ready` of PhaseManager into a new function and then call that function manually after the scene is loaded?

Reply:
I don't really see a problem with it in PhaseManager. 
We might not even need to touch phases at the start of this function and switching the phases at the end of it (by calling the function like current PhaseManager's ready) might be enough ?
 
I think the reason I did it like this was I wanted to switch to a Phase that doesn't have any events happening inside of it to prevent things happening while the scene is being cleaned and reloaded - although I'm not sure if that is actually necessary or even fulfills what I intended 🤔

Reply:
I'm thinking it may be sufficient to switch phases after the load is complete. However if we run into problems and want to prevent stuff from happening during the load process, we could potentially add a new phase like `GAME_END` or something

Reply:
`GAME_END` would be confusing, `SCENE_END` probably better
but anyway, when you load a new scene it will call the tree that will await for ready signals from it's children nodes, so I think it should all setup properly when reloading a new scene :thinking:

Reply:
What do we think the action here should be then ? 🤔 

Should I make the new "COMBAT_END / SCENE END / ..." phase, switch to it at the beginning of this and then call the equivalent of _ready function of phaseManager after the switch like @cheesefrycook suggested ? Or ?

Reply:
Well imo I think it should work without that
Meaning you shouldn't have to set the `GAME_STARTING` for the game to be in this specific phase, as I explained it should naturally be set like that when the scene starts

Reply:
PhaseManager is autoloaded, so just reloading the scene doesn't do anything with phases. 
If I remove the phase settings I am doing here, what happens is enemy turn finishes running (as scene reload is deferred) which switches phase to PLAYER_ATTACKING. After the reload, the phase stays on PLAYER_ATTACKING. The issue with this is that the cards are not served to the player (as cards are served when it detects change from some phase into PLAYER_ATTACKING) - game gets into a state where it is player's turn but he has no cards and can only end turn

Reply:
Aaaaah i forgot that PhaseManager is autoloaded, you are right ! So you indeed need to call the `_ready`
I think doing a `SCENE_END` phase is probably the best here then

Reply:
I wasn't sure if it would be better to do SCENE_END or call the _ready at the end (rather then just changing phase to player's turn) - in the end I did both. This way the SCENE_END is currently redundant I believe, but makes it more clear what's happening and might have use in the future ? [See here](https://github.com/Saplings-Projects/1M_sub/pull/60/files#diff-c6159dde6e380cb517c924628c3146eff48ca4e4b4d2cdc89af1c5adf9290290R47-R55)

Reply:
yes i read it, it's clear what happens; a good idea I think :+1: